### PR TITLE
Issue #1493098 by sun, jcisio, pcambra, Pol, wamilton, Rok Žlender, beejeebus: convert cron settings to configuration system.

### PR DESCRIPTION
**Pull request by hellomrcat**
PR for https://github.com/backdrop/backdrop-issues/issues/170

### CODE REVIEW

File: core/modules/system/system.admin.inc [hunk 1]
```diff
@@ -1482,11 +1483,29 @@ function system_cron_settings() {
...
     '#title' => t('Run cron every'),
     '#description' => t('More information about setting up scheduled tasks can be found by <a href="@url">reading the cron tutorial on drupal.org</a>.', array('@url' => url('http://drupal.org/cron'))),
-    '#default_value' => variable_get('cron_safe_threshold', BACKDROP_CRON_DEFAULT_THRESHOLD),
+    '#default_value' => config('system.cron')->get('cron_safe_threshold'),
     '#options' => array(0 => t('Never')) + backdrop_map_assoc(array(3600, 10800, 21600, 43200, 86400, 604800), 'format_interval'),
   );
+  $form_state['config']['cron_safe_threshold'] = array(
+    'name' => 'system.cron',
+    'path' => 'cron_safe_threshold',
+  );
```
Comment (quicksketch) [inline core/modules/system/system.admin.inc]: We can remove this chunk, as it doesn't do anything. Sorry the contact module example may have included it, but it should have been removed later (if it hasn't been already).

File: core/modules/system/system.module [hunk 2]
```diff
@@ -3711,7 +3711,7 @@ function system_run_automated_cron() {
   // If the site is not fully installed, suppress the automated cron run.
   // Otherwise it could be triggered prematurely by Ajax requests during
   // installation.
-  if (($threshold = variable_get('cron_safe_threshold', BACKDROP_CRON_DEFAULT_THRESHOLD)) > 0 && state_get('install_task') == 'done') {
+  if (($threshold = config('system.cron')->get('cron_safe_threshold')) > 0 && variable_get('install_task') == 'done') {
```
Comment (quicksketch) [inline core/modules/system/system.module]: state_get('install_task') should stay as state_get() and not variable_get().

File: core/modules/system/config/system.cron.json [hunk 3]
```diff
@@ -0,0 +1,7 @@
+{
+		"_config_name": "system.cron",
+		"cron_threshold_warning":	172800,
+		"cron_safe_threshold": 10800,
+		"cron_max_threshold": 10800,
+		"cron_threshold_error":	1209600
+}
```
Comment (quicksketch) [inline core/modules/system/config/system.cron.json:7]: This looks great but all tabs need to be spaces in this default file. Backdrop should actually format this file for you when you save the configuration for the first time.

File: core/includes/install.inc [hunk 4]
```diff
@@ -822,8 +823,18 @@ function backdrop_install_system() {
...
+  // system_rebuild_theme_data().
+  drupal_static_reset('system_list');
+  module_implements_reset();
+
   system_rebuild_module_data();
+  system_rebuild_theme_data();
+
   config_install_default_config('system');
+
+  module_invoke('system', 'install');
```
Comment (quicksketch) [inline core/includes/install.inc]: Do we need all this rearrangement in `backdrop_install_system()`? I don't think installing the cron settings should affect this function (but maybe it does?)

**Comment by quicksketch** [comment]
Looks like a great start! I left a bunch of inline comments on the changes.

File: core/includes/install.inc (hunk 4 above)
Comment (hellomrcat) [inline core/includes/install.inc]: @quicksketch I'm not sure if it's necessary, for this function I was just following [this patch](https://www.drupal.org/files/drupal8.config-cron.76.patch). I'll try taking it out and see if it works.

Comment (hellomrcat) [inline core/includes/install.inc]: According to sun on the drupal forums: 

> Apparently, the Drupal installer does not install system module in the regular/proper procedure. Due to that, the
> system.cron:key is never actually set/updated after the Drupal installation. Attached patch additionally fixes the
> Drupal installer to install System module correctly.

Comment (quicksketch) [inline core/includes/install.inc]: Ah, excellent. Well putting things in the right order sounds like a good idea. :)

File: core/modules/system/system.admin.inc [hunk 5]
```diff
@@ -1482,11 +1483,25 @@ function system_cron_settings() {
     '#type' => 'select',
     '#title' => t('Run cron every'),
     '#description' => t('More information about setting up scheduled tasks can be found by <a href="@url">reading the cron tutorial on drupal.org</a>.', array('@url' => url('http://drupal.org/cron'))),
-    '#default_value' => variable_get('cron_safe_threshold', BACKDROP_CRON_DEFAULT_THRESHOLD),
+    '#default_value' => config('system.cron')->get('cron_safe_threshold'),
     '#options' => array(0 => t('Never')) + backdrop_map_assoc(array(3600, 10800, 21600, 43200, 86400, 604800), 'format_interval'),
   );
 
-  return system_settings_form($form);
+  $form['submit'] = array('#type' => 'submit', '#value' => t('Save configuration'));
```
Comment (quicksketch) [inline core/modules/system/system.admin.inc]: To be consistent with other forms, let's change this to:

``` php
$form['actions']['#type'] = 'actions';
$form['actions']['submit'] = array(
  '#type' => 'submit',
  '#value' => t('Save configuration'),
);
```

File: core/modules/system/system.admin.inc [hunk 6]
```diff
@@ -1482,11 +1483,25 @@ function system_cron_settings() {
...
+  return $form;
+}
+
+/**
+ * Form builder submit handler; Handle submission for cron settings.
```
Comment (quicksketch) [inline core/modules/system/system.admin.inc]: Comment here could be "Form submit handler", rather than "Form builder submit handler".

**Comment by quicksketch** [comment]
This all looks great! Nice job. :D

A few more minor items are needed here:
- Let's delete the constant BACKDROP_CRON_DEFAULT_THRESHOLD entirely, it's not used any more.
- We need to add this new config file to system_config_info() so that it is exportable/diffable from the Configuration UI interface.

**Comment by quicksketch** [comment]
Looks great! Tested it out locally and the update works flawlessly. Code looks good. Merged in. Thanks @hellomrcat!

**Pull request merged**